lib/_http_server.js

```diff
@@ -151,38 +151,32 @@ ServerResponse.prototype._implicitHeader = function() {
   this.writeHead(this.statusCode);
 };
 
-ServerResponse.prototype.writeHead = function(statusCode, reason, obj) {
-  var headers;
-
-  if (typeof reason === 'string') {
-    // writeHead(statusCode, reasonPhrase[, headers])
-    this.statusMessage = reason;
+ServerResponse.prototype.writeHead = function(statusCode, statusMsg, headers) {
+  if (typeof statusMsg === 'string') {
+    // writeHead(statusCode, statusMessage[, headers])
+    this.statusMessage = statusMsg;
   } else {
     // writeHead(statusCode[, headers])
     this.statusMessage =
         this.statusMessage || STATUS_CODES[statusCode] || 'unknown';
-    obj = reason;
+    headers = statusMsg;
   }
   this.statusCode = statusCode;
 
   if (this._headers) {
     // Slow-case: when progressive API and header fields are passed.
-    if (obj) {
-      var keys = Object.keys(obj);
+    if (headers) {
+      var keys = Object.keys(headers);
       for (var i = 0; i < keys.length; i++) {
-        var k = keys[i];
-        if (k) this.setHeader(k, obj[k]);
+        var key = keys[i];
+        this.setHeader(key, headers[key]);
```

Comment:
This was the check I was worried about and wanted someone else's eyes on - it was introduced in fe83861, but I can't find the PR for it right now. It seemed pointless to me, and all tests pass without it, but I still want to figure out why it was there in the first place, just like you.

When I saw it, my first guess was a safeguard against something like this?:
```javascript
[ 'Content-Type', 'X-Foo': 'Bar' ]
```